### ISSUE TITLE
Cannot look up or complete external crates

### ISSUE THREAD

**Issue by medusacle** [racer-rust/racer#995]
I still seem to be having the issue in #551. I carefully followed the instructions to install `racer` and `vim-racer`. Versions:
```
racer 2.1.9
rustc 1.31.0-beta.1 (2824a67b0 2018-10-29)
rustc 1.31.0-nightly (1cf82fd9c 2018-10-30)
cargo 1.31.0-beta (efb7972a0 2018-10-24)
```

It wasn't working in vim so I tried manually. this works:
```bash
user@rust:~/game/roguelike$ racer complete std::io::
MATCH prelude,1,0,/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/prelude.rs,Module,/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/io/prelude.rs
...
```

this doesn't, even though these are local dependencies and specified in `Cargo.lock` (project was successfully built using `cargo build`):
```bash
user@rust:~/game/roguelike$ racer complete tcod::
user@rust:~/game/roguelike$ racer complete rand::
user@rust:~/game/roguelike$ grep rand Cargo.lock
name = "rand"
 "rand 0.4.3 (registry+https://github.com/rust-lang/crates.io-index)",
name = "rand"
 "rand 0.3.22 (registry+https://github.com/rust-lang/crates.io-index)",
"checksum rand 0.3.22 (registry+https://github.com/rust-lang/crates.io-index)" = "15a732abf9d20f0ad8eeb6f909bf6868722d9a06e1e50802b6a70351f40b4eb1"
"checksum rand 0.4.3 (registry+https://github.com/rust-lang/crates.io-index)" = "8356f47b32624fef5b3301c1be97e5944ecdd595409cc5da11d05f211db6cfbd"
```

with debugging flags
```bash
user@rust:~/game/roguelike$ RUST_BACKTRACE=1 RUST_LOG=racer=trace racer complete rand::
DEBUG 2018-11-03T07:52:43.696344656Z: racer::util: Getting rust source path. Trying env var RUST_SRC_PATH.
DEBUG 2018-11-03T07:52:43.696719282Z: racer::nameres: do_file_search with search string "rand"
DEBUG 2018-11-03T07:52:43.696928681Z: racer::util: Getting rust source path. Trying env var RUST_SRC_PATH.
DEBUG 2018-11-03T07:52:43.697154398Z: racer::nameres: do_file_search std_path: Some("/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src")
DEBUG 2018-11-03T07:52:43.697387775Z: racer::nameres: do_file_search v: ["/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src", "."]
```
also tried lookup used on path/line number: no luck !

I've tried stracing recursively to see if it loads `Cargo.lock` at all but looks like it doesn't, doesn't look like it's trying at all:
```bash
user@rust:~/game/roguelike$ strace -f racer complete rand:: 2>&1 | grep Cargo
(nothing)
```
(`cargo` is in the system path, and so is `racer` itself)

…what am I missing?

**Comment by kngwyu** [comment]
Looks same as #981 
Could you please try current master?
If it works I'll publish it.

**Comment by medusacle** [comment]
thank you! just tried,
```bash
user@rust:~/upstream$ cargo +nightly uninstall racer
    Removing /home/user/.cargo/bin/racer
user@rust:~/upstream$ git clone https://github.com/racer-rust/racer.git
user@rust:~/upstream/racer$ cd racer
user@rust:~/upstream/racer$ cargo +nightly build --release
user@rust:~/upstream/racer$ mv target/release/racer ~/.cargo/bin/
user@rust:~/upstream/racer$ racer --version
racer 2.1.9
```
Version still shows as 2.1.9—but I suppose that's expected.

- unfortunately, it didn't help for the `complete <ident>` case:
```
user@rust:~/game/roguelike$ RUST_BACKTRACE=1 RUST_LOG=racer=trace racer complete rand::
DEBUG 2018-11-03T09:59:49.770705276Z: racer::util: Getting rust source path. Trying env var RUST_SRC_PATH.
DEBUG 2018-11-03T09:59:49.770803377Z: racer::nameres: do_file_search with search string "rand"
DEBUG 2018-11-03T09:59:49.770820477Z: racer::util: Getting rust source path. Trying env var RUST_SRC_PATH.
DEBUG 2018-11-03T09:59:49.770851736Z: racer::nameres: do_file_search std_path: Some("/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src")
DEBUG 2018-11-03T09:59:49.770903497Z: racer::nameres: do_file_search v: ["/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src", "."]
```

- interestingly, completion by line/col now does work ! this is likely enough to get it to work in vim,
```bash
user@rust:~/game/roguelike$ racer complete 496 18 src/main.rs
PREFIX 19500,19502,wi
MATCH window_closed,297,11,/home/user/game/tcod-rs/src/console.rs,Function,pub fn window_closed(&self) -> bool
MATCH width,680,7,/home/user/game/tcod-rs/src/console.rs,Function,fn width(&self) -> i32
END
```
```bash
user@rust:~/game/roguelike$ racer complete 314 24 src/main.rs
PREFIX 12060,12063,ran
MATCH rand4,1,0,/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.4.3/src/lib.rs,Module,/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.4.3/src/lib.rs
MATCH random,955,7,/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.4.3/src/lib.rs,Function,pub fn random<T: Rand>() -> T
END
```

**Comment by juchiast** [comment]
It would be helpful to run "cargo test" in racer, just to verify that everything works fine. If some test fails, you get a starting point.

**Comment by kngwyu** [comment]
@juchiast
This bug was hidden by travis's cache and I think it's an exceptional case.

**Comment by medusacle** [comment]
@juchiast various path related tests are failing here:
```
test util::test_get_rust_src_path_env_ok ... FAILED
test util::test_get_rust_src_path_missing ... FAILED
test util::test_get_rust_src_path_rustup_ok ... FAILED
test util::test_get_rust_src_path_not_rust_source_tree ... FAILED
test nameres::test_do_file_search_local ... ok
test nameres::test_do_file_search_std ... FAILED
...
failures:

---- util::test_get_rust_src_path_env_ok stdout ----
thread 'util::test_get_rust_src_path_env_ok' panicked at 'called `Option::unwrap()` on a `None` value', libcore/option.rs:355:21
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- util::test_get_rust_src_path_missing stdout ----
thread 'util::test_get_rust_src_path_missing' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', libcore/result.rs:1009:5

---- util::test_get_rust_src_path_rustup_ok stdout ----
thread 'util::test_get_rust_src_path_rustup_ok' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', libcore/result.rs:1009:5

---- util::test_get_rust_src_path_not_rust_source_tree stdout ----
thread 'util::test_get_rust_src_path_not_rust_source_tree' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', libcore/result.rs:1009:5

---- nameres::test_do_file_search_std stdout ----
thread 'nameres::test_do_file_search_std' panicked at 'assertion failed: matches.into_iter().any(|m| m.filepath.ends_with("src/libstd/lib.rs"))', src/racer/nameres.rs:763:5
```

**Comment by kngwyu** [comment]
These tests are related to rust-src but I think you don't have to mind if everything work fine.
What's your problem now?

**Comment by medusacle** [comment]
> It's intended and I don't know what's this feature for.

yes if that particular way is not important that's fine with me :woman_shrugging: everything is OK then

**Comment by kngwyu** [comment]
Published 2.1.10 now

**Comment by nilly** [comment]
racer  2.1.10  also  has  this  issue. 

can not  complete   third  party  crate
